coding/bwt_coder.hpp

```diff
@@ -0,0 +1,117 @@
+#pragma once
+
+#include "coding/huffman.hpp"
+#include "coding/varint.hpp"
+
+#include "base/bwt.hpp"
+#include "base/move_to_front.hpp"
+
+#include <algorithm>
+#include <cstdint>
+#include <iterator>
+#include <vector>
+
+namespace coding
+{
+class BWTCoder
+{
+public:
+  struct Params
+  {
+    size_t m_blockSize = 32000;
+  };
+
+  template <typename Sink>
+  static void EncodeAndWriteBlock(Sink & sink, size_t n, uint8_t const * s,
+                                  std::vector<uint8_t> & bwtBuffer)
+  {
+    bwtBuffer.resize(n);
+    auto const start = base::BWT(n, s, bwtBuffer.data());
+
+    base::MoveToFront mtf;
+    for (auto & b : bwtBuffer)
+      b = mtf.Transform(b);
+
+    WriteVarUint(sink, start);
+
+    HuffmanCoder huffman;
+    huffman.Init(bwtBuffer.begin(), bwtBuffer.end());
+    huffman.WriteEncoding(sink);
+    huffman.EncodeAndWrite(sink, bwtBuffer.begin(), bwtBuffer.end());
+  }
+
+  template <typename Sink>
+  static void EncodeAndWriteBlock(Sink & sink, size_t n, uint8_t const * s)
+  {
+    vector<uint8_t> bwtBuffer;
+    EncodeAndWriteBlock(sink, n, s, bwtBuffer);
+  }
+
+  template <typename Sink>
+  static void EncodeAndWrite(Params const & params, Sink & sink, size_t n, uint8_t const * s)
+  {
+    CHECK(params.m_blockSize != 0, ());
+    CHECK_GREATER(n + params.m_blockSize, n, ());
```

Comment:
second CHECK checks params.m_blocksSize != 0, then it is enough only second check :)